src/modules/xworkspaces.cpp

```diff
@@ -124,27 +129,38 @@ namespace modules {
    */
   void xworkspaces_module::rebuild_clientlist() {
     vector<xcb_window_t> clients = ewmh_util::get_client_list();
-    vector<xcb_window_t> diff;
+    vector<xcb_window_t> new_windows;
+
     std::sort(clients.begin(), clients.end());
     std::sort(m_clientlist.begin(), m_clientlist.end());
-
-    if (m_clientlist.size() > clients.size()) {
-      std::set_difference(
-          m_clientlist.begin(), m_clientlist.end(), clients.begin(), clients.end(), back_inserter(diff));
-      for (auto&& win : diff) {
-        // untrack window
-        m_clientlist.erase(std::remove(m_clientlist.begin(), m_clientlist.end(), win), m_clientlist.end());
+    std::fill(m_desktop_occupied.begin(), m_desktop_occupied.end(), false);
+
+    auto it_old = clients.begin();
+    auto it_new = m_clientlist.begin();
+
+    // Traverse both the stored and the up-to-date clientlist to update our
+    // stored list with as little changes as possible (add windows that are
+    // only in the new list, remove windows that are no longer in the list)
+    while (it_old != clients.end() || it_new != m_clientlist.end()) {
+      if (it_old != clients.end()) {
+        auto desktop = ewmh_util::get_desktop_from_window(*it_old);
+        m_desktop_occupied[desktop] = true;
```

Comment:
```suggestion
        if (desktop != 0xffffffff) {
          m_desktop_occupied[desktop] = true;
        }
```
Fixes a crash with windows with no desktop. (like i3 sticky windows, …)
See https://github.com/polybar/polybar/pull/882#issuecomment-483458242 and https://github.com/polybar/polybar/pull/882#issuecomment-483488566

Reply:
This would fix the crash, but would it handle `0xFFFFFFFF` correctly? If it should appear on all desktops, I would somehow expect all desktops to be occupied. Or maybe none, because it is so omnipresent that it does not need special highlighting. :confused:

Reply:
`0xFFFFFFFF` isn't a valid id for a desktop.
For example `polybar` is a window that should appears on every desktop.
I think we shouldn't consider that all desktop are occupied.
In i3wm, you have the concept of sticky windows that are above the desktop, so they aren't on a desktop, there are really above them.